includes/class-llms-dom-document.php

```diff
@@ -0,0 +1,160 @@
+<?php
+/**
+ * A convenient wrapper for the DOMDocument Class
+ *
+ * @package LifterLMS/Classes
+ *
+ * @since [version]
+ * @version [version]
+ */
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * LLMS_DOM_Document Class
+ *
+ * @since [version]
+ */
+class LLMS_DOM_Document {
+
+	/**
+	 * Stores the load method name
+	 *
+	 * @var string
+	 */
+	private $load_method = 'load_with_mb_convert_encoding';
+
+	/**
+	 * Stores the HTML string to load
+	 *
+	 * @var string
+	 */
+	private $source;
+
+	/**
+	 * Stores the DOMDocument instance
+	 *
+	 * @var DOMDocument
+	 */
+	private $dom;
+
+	/**
+	 * Stores loading errors
+	 *
+	 * @var null|WP_Error
+	 */
+	private $error;
+
+	/**
+	 * This forces DOMDocument to convert non-utf8 characters into HTML entities and without relying on `mb_convert_encoding()`.
+	 *
+	 * @var string
+	 */
+	private $utf8_fixer = '<meta id="llms-get-dom-doc-utf-fixer" http-equiv="Content-Type" content="text/html; charset=utf-8">';
+
+	/**
+	 * Constructor
+	 *
+	 * @since [version]
+	 *
+	 * @param string $source An HTML string, either a full HTML document or a partial string.
+	 * @return void|WP_Error
+	 */
+	public function __construct( $source ) {
+
+		if ( ! class_exists( 'DOMDocument' ) ) {
+			return new WP_Error( 'llms-dom-document-missing', __( 'DOMDocument not available.', 'lifterlms' ) );
+		}
+
+		/**
+		 * Filters the convert encoding method to be used when loading the source in the DOMDocument
+		 *
+		 * @param boolean $use_mb_convert_endoding Whether or not the convert encoding method should be used when loading the source in the DOMDocument.
+		 *                                         Default is `true`. Requires `mbstring` PHP extension.
+		 */
+		$use_mb_convert_encoding = apply_filters( 'llms_dom_document_use_mb_convert_encoding', true );
+		if ( ! ( $use_mb_convert_encoding && function_exists( 'mb_convert_encoding' ) ) ) {
+			$this->load_method = 'load_with_meta_utf_fixer';
+		}
+
+		$this->source = $source;
+		$this->dom    = new DOMDocument();
+	}
+
+	/**
+	 * Load the HTML string in the DOMDocument
+	 *
+	 * This function suppresses PHP warnings that would be thrown by DOMDocument when
+	 * loading a partial string or an HTML string with errors.
+	 *
+	 * @since [version]
+	 *
+	 * @return boolean|WP_Error Returns true if the source is loaded fine.
+	 *                          Or an error object when an error is encountered during loading.
+	 */
+	public function load() {
+
```

Comment:
We should check here if we already have an error and return early with the error (found in the constructor) if we do

Reply:
yeah, done in 47d61de